floss/language/go/extract.py

```diff
@@ -0,0 +1,371 @@
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+
+import re
+import sys
+import struct
+import logging
+import pathlib
+import argparse
+from typing import List, Iterable, Optional
+from itertools import chain
+
+import pefile
+
+from floss.results import StaticString, StringEncoding
+
+logger = logging.getLogger(__name__)
+
+MIN_STR_LEN = 6
+
+
+def yield_string(binary_string, addr, min_length):
+    try:
+        decoded_string = binary_string.decode("utf-8")
+        if decoded_string.isprintable() and len(decoded_string) >= min_length:
+            yield StaticString(string=decoded_string, offset=addr, encoding=StringEncoding.UTF8)
+    except UnicodeDecodeError:
+        pass
+
+
+def extract_strings_from_import_data(pe: pefile.PE) -> Iterable[StaticString]:
+    """Extract strings from the import data"""
+
+    for entry in pe.DIRECTORY_ENTRY_IMPORT:
+        for imp in entry.imports:
+            if imp.name is not None:
+                yield StaticString(string=imp.name.decode("utf-8"), offset=imp.address, encoding=StringEncoding.UTF8)
+
+
+def extract_build_id(section_data, min_length) -> Iterable[StaticString]:
+    # Build ID is a string that starts with "\xff\x20 Go build ID: " and ends with "\n"
+    # FF 20 47 6F 20 62 75 69  6C 64 20 49 44 3A 20 22  . Go build ID: "
+    # 36 4E 31 4D 77 6E 30 31  72 46 6E 41 51 4B 62 5A  6N1Mwn01rFnAQKbZ
+    # 73 46 5A 32 2F 38 41 6B  75 63 4B 46 58 4D 49 54  sFZ2/8AkucKFXMIT
+    # 63 51 52 49 75 55 5F 79  32 2F 62 76 4F 67 56 37  cQRIuU_y2/bvOgV7
+    # 52 4D 54 72 77 73 7A 5A  39 57 7A 69 6C 64 2F 72  RMTrwszZ9Wzild/r
+    # 33 31 50 47 70 61 6B 2D  48 77 36 4B 72 77 59 6E  31PGpak-Hw6KrwYn
+    # 52 4E 73 22 0A 20 FF CC  CC CC CC CC CC CC CC CC  RNs". ..........
+
+    build_id_regex = re.compile(b"(?<=\xff\x20)(.)*\x0A")
+
+    for m in build_id_regex.finditer(section_data):
+        addr = m.start()
+        binary_string = m.group(0)[0:-1]
+        yield from yield_string(binary_string, addr, min_length)
+        break
+
+
+def extract_stackstring(extract_stackstring_pattern, section_data, min_length) -> Iterable[StaticString]:
+    for m in extract_stackstring_pattern.finditer(section_data):
+        for i in range(1, 8):
+            try:
+                tmp_string = m.group(i)
+                if tmp_string != b"":
+                    addr = m.start()
+                    yield from yield_string(tmp_string, addr, min_length)
+            except AttributeError:
+                pass
+
+
+def extract_string_blob(section_data, min_length) -> Iterable[StaticString]:
+    # Extract string blob in .rdata section
+    """
+    0048E620  5B 34 5D 75 69 6E 74 38  00 09 2A 5B 38 5D 69 6E  [4]uint8..*[8]in
```

Comment:
these strings appear to be related to RTTI / reflection. two things suggest this:

  1. their contents, which are Go type names, and
  2. working backwards, to see where they're used, are type descriptors (`rtype` and friends).

it might be possible to reconstruct these type descriptors, such as structures and their fields, used in a Go program, and then we'd be able to link the string to the type/structure/metadata.

the upside: more complete information
downside: more work

the existing approach is simple and probably works fine. maybe we can update the documentation here to clarify what strings these likely are (reflection-related).

Reply:
parsing type name strings, like we see here, is described: https://github.com/mandiant/GoReSym/blob/fd518058f5d2998ebbc28d317ab5d29b4f34ae91/objfile/objfile.go#L800

notably, the leading byte uses 3 bits (least significant) to encode information, so the regex should take this into account.